File: solutions/misc/frequently-asked-questions/binary-search-problems.js
/**
 * Binary Search — 4 Core Problems
 *
 * Template:
 *   lo = 0, hi = n-1
 *   while lo <= hi:
 *     mid = (lo + hi) >> 1
 *     if check(mid): answer = mid; hi = mid-1  (or lo = mid+1)
 *     else: lo = mid+1  (or hi = mid-1)
 *
 * Problems:
 *   LC 33  — Search in Rotated Sorted Array (Medium)
 *   LC 153 — Find Minimum in Rotated Sorted Array (Medium)
 *   LC 162 — Find Peak Element (Medium)
 *   LC 704 — Binary Search (Easy)
 */

// LC 704 — Binary Search — O(log n)
function search(nums, target) {
  let lo = 0, hi = nums.length - 1;
  while (lo <= hi) {
    const mid = (lo + hi) >> 1;
    if (nums[mid] === target) return mid;
    if (nums[mid] < target) lo = mid + 1;
    else hi = mid - 1;
  }
  return -1;
}

// LC 33 — Search in Rotated Sorted Array — O(log n)
function searchRotated(nums, target) {
  let lo = 0, hi = nums.length - 1;
  while (lo <= hi) {
    const mid = (lo + hi) >> 1;
    if (nums[mid] === target) return mid;

    // Left half is sorted
    if (nums[lo] <= nums[mid]) {
      if (nums[lo] <= target && target < nums[mid]) hi = mid - 1;
      else lo = mid + 1;
    } else {
      // Right half is sorted
      if (nums[mid] < target && target <= nums[hi]) lo = mid + 1;
      else hi = mid - 1;
    }
  }
  return -1;
}

// LC 153 — Find Minimum in Rotated Sorted Array — O(log n)
function findMin(nums) {
  let lo = 0, hi = nums.length - 1;
  while (lo < hi) {
    const mid = (lo + hi) >> 1;
    if (nums[mid] > nums[hi]) lo = mid + 1; // min is in right half
    else hi = mid;                           // min is at mid or left
  }
  return nums[lo];
}

// LC 162 — Find Peak Element — O(log n)
// A peak is where nums[i] > nums[i-1] and nums[i] > nums[i+1]
function findPeakElement(nums) {
  let lo = 0, hi = nums.length - 1;
  while (lo < hi) {
    const mid = (lo + hi) >> 1;
    if (nums[mid] > nums[mid + 1]) hi = mid;     // peak is at mid or left
    else lo = mid + 1;                            // peak is right
  }
  return lo;
}

// ─── Tests ───────────────────────────────────────────────────────────────────
console.log(search([1,2,3,4,5,6], 4));              // 3
console.log(searchRotated([4,5,6,7,0,1,2], 0));     // 4
console.log(searchRotated([4,5,6,7,0,1,2], 3));     // -1
console.log(findMin([3,4,5,1,2]));                  // 1
console.log(findMin([4,5,6,7,0,1,2]));              // 0
console.log(findPeakElement([1,2,3,1]));            // 2

module.exports = { search, searchRotated, findMin, findPeakElement };

File: solutions/misc/frequently-asked-questions/binary-search-problems.py
"""
Binary Search — 4 Core Problems (Python)

LC 33  — Search in Rotated Sorted Array
LC 153 — Find Minimum in Rotated Sorted Array
LC 162 — Find Peak Element
LC 704 — Binary Search
"""

from typing import List


# LC 704 — O(log n)
def search(nums: List[int], target: int) -> int:
    """Standard binary search on sorted array."""
    lo, hi = 0, len(nums) - 1
    while lo <= hi:
        mid = (lo + hi) >> 1  # Bitwise divide by 2 (equivalent to (lo+hi)//2)
        if nums[mid] == target: return mid
        elif nums[mid] < target: lo = mid + 1  # Target in right half
        else: hi = mid - 1  # Target in left half
    return -1


# LC 33 — O(log n)
def search_rotated(nums: List[int], target: int) -> int:
    """Search in rotated sorted array (no duplicates)."""
    lo, hi = 0, len(nums) - 1
    while lo <= hi:
        mid = (lo + hi) >> 1
        if nums[mid] == target: return mid
        # Determine which half is sorted
        if nums[lo] <= nums[mid]:  # Left half is sorted
            if nums[lo] <= target < nums[mid]: hi = mid - 1  # Target in left half
            else: lo = mid + 1  # Target in right half
        else:  # Right half is sorted
            if nums[mid] < target <= nums[hi]: lo = mid + 1  # Target in right half
            else: hi = mid - 1  # Target in left half
    return -1


# LC 153 — O(log n)
def find_min(nums: List[int]) -> int:
    """Find minimum in rotated sorted array (no duplicates)."""
    lo, hi = 0, len(nums) - 1
    while lo < hi:
        mid = (lo + hi) >> 1
        # If mid > hi, min is in right half (rotation point)
        if nums[mid] > nums[hi]: lo = mid + 1
        else: hi = mid  # Min is in left half (including mid)
    return nums[lo]


# LC 162 — O(log n)
def find_peak_element(nums: List[int]) -> int:
    """Find a peak element (nums[i] > nums[i-1] and nums[i] > nums[i+1])."""
    lo, hi = 0, len(nums) - 1
    while lo < hi:
        mid = (lo + hi) >> 1
        # If mid > mid+1, we're on descending slope → peak is to the left
        if nums[mid] > nums[mid + 1]: hi = mid
        # If mid < mid+1, we're on ascending slope → peak is to the right
        else: lo = mid + 1
    return lo


# ── Tests ─────────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    print(search([1,2,3,4,5,6], 4))              # 3
    print(search_rotated([4,5,6,7,0,1,2], 0))    # 4
    print(search_rotated([4,5,6,7,0,1,2], 3))    # -1
    print(find_min([3,4,5,1,2]))                 # 1
    print(find_min([4,5,6,7,0,1,2]))             # 0
    print(find_peak_element([1,2,3,1]))          # 2
